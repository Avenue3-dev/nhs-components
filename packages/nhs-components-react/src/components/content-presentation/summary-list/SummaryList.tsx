import clsx from 'clsx'
import { HTMLProps } from 'react'

type SummaryListProps = {
  borderless?: boolean
} & HTMLProps<HTMLDListElement>

type SummaryList = {
  Row: typeof Row
  Key: typeof Key
  Value: typeof Value
  Actions: typeof Actions
} & React.FC<SummaryListProps>

const Row: React.FC<HTMLProps<HTMLDivElement>> = ({
  children,
  className,
  ...rest
}): JSX.Element => {
  return (
    <div className={clsx('nhsuk-summary-list__row', className)} {...rest}>
      {children}
    </div>
  )
}

const Key: React.FC<HTMLProps<HTMLDivElement>> = ({
  children,
  className,
  ...rest
}): JSX.Element => {
  return (
    <dt className={clsx('nhsuk-summary-list__key', className)} {...rest}>
      {children}
    </dt>
  )
}

const Value: React.FC<HTMLProps<HTMLDivElement>> = ({
  children,
  className,
  ...rest
}): JSX.Element => {
  return (
    <dd className={clsx('nhsuk-summary-list__value', className)} {...rest}>
      {children}
    </dd>
  )
}

const Actions: React.FC<HTMLProps<HTMLDivElement>> = ({
  children,
  className,
  ...rest
}): JSX.Element => {
  return (
    <dd className={clsx('nhsuk-summary-list__actions', className)} {...rest}>
      {children}
    </dd>
  )
}

const SummaryList: SummaryList = ({
  children,
  className,
  borderless,
  ...rest
}): JSX.Element => {
  return (
    <dl
      className={clsx(
        'nhsuk-summary-list',
        { 'nhsuk-summary-list--no-border': borderless },
        className,
      )}
      {...rest}
    >
      {children}
    </dl>
  )
}

SummaryList.displayName = 'SummaryList'
Row.displayName = 'SummaryList.Row'
Key.displayName = 'SummaryList.Key'
Value.displayName = 'SummaryList.Value'
Actions.displayName = 'SummaryList.Actions'

SummaryList.Row = Row
SummaryList.Key = Key
SummaryList.Value = Value
SummaryList.Actions = Actions

export default SummaryList
