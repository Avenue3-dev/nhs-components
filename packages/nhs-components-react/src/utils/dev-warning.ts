const prod = process.env.NODE_ENV === 'production'

export const experimentalWarning = (
  component: 'tabs' | 'character-count',
): void => {
  if (prod) {
    return
  }

  console.warn(`ATTENTION
  This component is tagged as experimental and therefore could change without notice. See more details about this component at https://service-manual.nhs.uk/design-system/components/${component}
  `)
}
