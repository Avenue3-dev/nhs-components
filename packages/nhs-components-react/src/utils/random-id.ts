export const generateRandomString = (length = 5): string => {
  const rand = Math.random() + 1

  return rand.toString(36).substring(2, length + 2)
}

export const generateRandomName = (prefix?: string): string => {
  const randString = generateRandomString()

  return prefix ? `${prefix}_${randString}` : randString
}

export const generateRandomId = generateRandomName
