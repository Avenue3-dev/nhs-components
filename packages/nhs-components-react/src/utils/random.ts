export const generateRandomString = (length: number = 8): string => {
  return [...Array<string>(length)]
    .map(() => Math.random().toString(36)[2])
    .join('')
}
