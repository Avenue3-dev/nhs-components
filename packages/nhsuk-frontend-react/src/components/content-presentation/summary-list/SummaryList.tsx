import { Base } from '@/components/core/base/Base';
import { VisuallyHidden } from '@/components/core/visually-hidden/VisuallyHidden';
import { Factory, factory } from '@/internal/factory/factory';
import {
  PolymorphicFactory,
  polymorphicFactory,
} from '@/internal/factory/polymorphic-factory';
import { AsElementProps, ElementProps } from '@/types/shared';
import clsx from 'clsx';
import React from 'react';

export type SummaryListProps = ElementProps<'dl'>;

type SummaryListFactory = Factory<{
  props: SummaryListProps;
  ref: HTMLDListElement;
  staticComponents: {
    Row: typeof SummaryListRow;
  };
}>;

const SummaryList = factory<SummaryListFactory>(
  ({ className, ...props }, ref) => {
    return (
      <dl
        className={clsx('nhsuk-summary-list', className)}
        {...props}
        ref={ref}
      />
    );
  },
);

export type SummaryListRowProps = ElementProps<'div'>;

type SummaryListRowFactory = Factory<{
  props: SummaryListRowProps;
  ref: HTMLDivElement;
  staticComponents: {
    Key: typeof SummaryListRowKey;
    Value: typeof SummaryListRowValue;
    Actions: typeof SummaryListRowActions;
    ActionLink: typeof SummaryListRowActionLink;
  };
}>;

const SummaryListRow = factory<SummaryListRowFactory>(
  ({ className, ...props }, ref) => {
    return (
      <div
        className={clsx('nhsuk-summary-list__row', className)}
        {...props}
        ref={ref}
      />
    );
  },
);

export type SummaryListRowKeyProps = ElementProps<'dt'>;

const SummaryListRowKey = ({ className, ...props }: SummaryListRowKeyProps) => {
  return (
    <dt className={clsx('nhsuk-summary-list__key', className)} {...props} />
  );
};

export type SummaryListRowValueProps = ElementProps<'dd'>;

const SummaryListRowValue = ({
  className,
  ...props
}: SummaryListRowValueProps) => {
  return (
    <dd className={clsx('nhsuk-summary-list__value', className)} {...props} />
  );
};

export type SummaryListRowActionsProps = ElementProps<'dd'>;

const SummaryListRowActions = ({
  className,
  ...props
}: SummaryListRowActionsProps) => {
  return (
    <dd className={clsx('nhsuk-summary-list__actions', className)} {...props} />
  );
};

export type SummaryListRowActionLinkProps = {
  visuallyHiddenText: string;
} & ElementProps<'a'>;

type SummaryListRowActionLinkFactory = PolymorphicFactory<{
  props: SummaryListRowActionLinkProps;
  defaultComponent: 'a';
  defaultRef: HTMLAnchorElement;
}>;

const SummaryListRowActionLink =
  polymorphicFactory<SummaryListRowActionLinkFactory>(
    (
      {
        children,
        as: component = 'a',
        visuallyHiddenText,
        ...props
      }: SummaryListRowActionLinkProps & AsElementProps,
      ref,
    ) => {
      return (
        <Base as={component} ref={ref} {...props}>
          {children}
          <VisuallyHidden> {visuallyHiddenText}</VisuallyHidden>
        </Base>
      );
    },
  );

SummaryList.displayName = 'SummaryList';
SummaryListRow.displayName = 'SummaryList.Row';
SummaryListRowKey.displayName = 'SummaryList.Row.Key';
SummaryListRowValue.displayName = 'SummaryList.Row.Value';
SummaryListRowActions.displayName = 'SummaryList.Row.Actions';
SummaryListRowActionLink.displayName = 'SummaryList.Row.ActionLink';

SummaryListRow.Key = SummaryListRowKey;
SummaryListRow.Value = SummaryListRowValue;
SummaryListRow.Actions = SummaryListRowActions;
SummaryListRow.ActionLink = SummaryListRowActionLink;

SummaryList.Row = SummaryListRow;

export {
  SummaryList,
  SummaryListRow,
  SummaryListRowKey,
  SummaryListRowValue,
  SummaryListRowActions,
  SummaryListRowActionLink,
};
