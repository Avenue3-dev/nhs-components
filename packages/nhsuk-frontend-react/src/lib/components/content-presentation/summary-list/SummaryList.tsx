import clsx from 'clsx'
import { HTMLProps } from 'react'

type SummaryListProps = {
  borderless?: boolean
} & HTMLProps<HTMLDListElement>

type SummaryList = {
  Row: typeof Row
  Key: typeof Key
  Value: typeof Value
  Actions: typeof Actions
} & React.FC<SummaryListProps>

const Row: React.FC<HTMLProps<HTMLDivElement>> = ({
  children,
  className,
  ...rest
}): JSX.Element => {
  return (
    <div className={clsx('nhsuk-summary-list__row', className)} {...rest}>
      {children}
    </div>
  )
}

const Key: React.FC<HTMLProps<HTMLDivElement>> = ({
  children,
  className,
  ...rest
}): JSX.Element => {
  return (
    <dt className={clsx('nhsuk-summary-list__key', className)} {...rest}>
      {children}
    </dt>
  )
}

const Value: React.FC<HTMLProps<HTMLDivElement>> = ({
  children,
  className,
  ...rest
}): JSX.Element => {
  return (
    <dd className={clsx('nhsuk-summary-list__value', className)} {...rest}>
      {children}
    </dd>
  )
}

const Actions: React.FC<HTMLProps<HTMLDivElement>> = ({
  children,
  className,
  ...rest
}): JSX.Element => {
  return (
    <dd className={clsx('nhsuk-summary-list__actions', className)} {...rest}>
      {children}
    </dd>
  )
}

/**
 * Use the summary list to summarise information, for example, a userâ€™s responses at the end of a form.
 *
 * For more information on when to use this component, go to the [NHS Service Manual website](https://service-manual.nhs.uk/design-system/components/summary-list).
 *
 * @link https://service-manual.nhs.uk/design-system/components/summary-list
 *
 * @example
 *
 * ```tsx
 * <SummaryList>
 *  <SummaryList.Row>
 *   <SummaryList.Key>Name</SummaryList.Key>
 *   <SummaryList.Value>Dr John Smith</SummaryList.Value>
 *   <SummaryList.Actions>
 *    <a href="#">Change<span className="nhsuk-u-visually-hidden"> name</span></a>
 *   </SummaryList.Actions>
 *  </SummaryList.Row>
 * </SummaryList>
 * ```
 */
const SummaryList: SummaryList = ({
  children,
  className,
  borderless,
  ...rest
}): JSX.Element => {
  return (
    <dl
      className={clsx(
        'nhsuk-summary-list',
        { 'nhsuk-summary-list--no-border': borderless },
        className,
      )}
      {...rest}
    >
      {children}
    </dl>
  )
}

SummaryList.displayName = 'SummaryList'
Row.displayName = 'SummaryList.Row'
Key.displayName = 'SummaryList.Key'
Value.displayName = 'SummaryList.Value'
Actions.displayName = 'SummaryList.Actions'

SummaryList.Row = Row
SummaryList.Key = Key
SummaryList.Value = Value
SummaryList.Actions = Actions

export default SummaryList
