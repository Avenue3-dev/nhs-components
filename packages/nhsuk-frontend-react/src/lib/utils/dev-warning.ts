const prod = process.env.NODE_ENV === 'production'

/**
 * Print a warning when using experimental components when not in production mode.
 *
 * @param component name of the experimental component.
 */
export const useExperimentalWarning = (
  component: 'tabs' | 'character-count',
): void => {
  useDevWarning(`This component is tagged as experimental and therefore could change without notice.

  See more details about this component at https://service-manual.nhs.uk/design-system/components/${component}
  `)
}

/**
 * Print a warning message when not in production mode.
 *
 * @param message message to print when not it production.
 */
export const useDevWarning = (message: string): void => {
  if (prod) {
    return
  }

  console.warn(message)
}
