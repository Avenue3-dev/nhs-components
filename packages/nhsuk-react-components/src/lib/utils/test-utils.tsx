/* eslint-disable import/export */
import {
  RenderOptions,
  RenderResult,
  cleanup,
  render,
} from '@testing-library/react'
import { MemoryRouter as Router } from 'react-router-dom'
import { afterEach } from 'vitest'

afterEach(() => {
  cleanup()
})

const customRender = (
  ui: React.ReactElement,
  options: RenderOptions = {},
): RenderResult => {
  return render(ui, {
    // wrap provider(s) here if needed
    wrapper: ({ children }) => <Router>{children}</Router>,
    ...options,
  })
}

export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'
// override render export
export { customRender as render }
