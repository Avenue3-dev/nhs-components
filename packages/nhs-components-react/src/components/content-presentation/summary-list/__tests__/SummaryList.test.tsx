import { render, screen } from '@/utils/test-utils'
import { describe, expect, test } from 'vitest'
import SummaryList from '..'

describe('SummaryList', () => {
  test('should match snapshot', () => {
    render(
      <SummaryList data-testid="summary-list">
        <SummaryList.Row>
          <SummaryList.Key>Label</SummaryList.Key>
          <SummaryList.Value>Value</SummaryList.Value>
          <SummaryList.Actions>
            <a href="#/change">Change</a>
          </SummaryList.Actions>
        </SummaryList.Row>
        <SummaryList.Row>
          <SummaryList.Key>Label</SummaryList.Key>
          <SummaryList.Value>Value</SummaryList.Value>
          <SummaryList.Actions>
            <a href="#/change">Change</a>
          </SummaryList.Actions>
        </SummaryList.Row>
        <SummaryList.Row>
          <SummaryList.Key>Label</SummaryList.Key>
          <SummaryList.Value>Value</SummaryList.Value>
          <SummaryList.Actions>
            <a href="#/change">Change</a>
          </SummaryList.Actions>
        </SummaryList.Row>
      </SummaryList>,
    )

    expect(screen.getByTestId('summary-list')).toMatchSnapshot()
  })

  test('should match snapshot - borderless', () => {
    render(
      <SummaryList data-testid="summary-list" borderless>
        <SummaryList.Row>
          <SummaryList.Key>Label</SummaryList.Key>
          <SummaryList.Value>Value</SummaryList.Value>
        </SummaryList.Row>
        <SummaryList.Row>
          <SummaryList.Key>Label</SummaryList.Key>
          <SummaryList.Value>Value</SummaryList.Value>
        </SummaryList.Row>
        <SummaryList.Row>
          <SummaryList.Key>Label</SummaryList.Key>
          <SummaryList.Value>Value</SummaryList.Value>
        </SummaryList.Row>
      </SummaryList>,
    )

    expect(screen.getByTestId('summary-list')).toMatchSnapshot()
  })
})
