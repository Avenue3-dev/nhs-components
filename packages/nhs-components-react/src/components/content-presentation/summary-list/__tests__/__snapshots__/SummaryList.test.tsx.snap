// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SummaryList > should match snapshot - borderless 1`] = `
<dl
  class="nhsuk-summary-list nhsuk-summary-list--no-border"
  data-testid="summary-list"
>
  <div
    class="nhsuk-summary-list__row"
  >
    <dt
      class="nhsuk-summary-list__key"
    >
      Label
    </dt>
    <dd
      class="nhsuk-summary-list__value"
    >
      Value
    </dd>
  </div>
  <div
    class="nhsuk-summary-list__row"
  >
    <dt
      class="nhsuk-summary-list__key"
    >
      Label
    </dt>
    <dd
      class="nhsuk-summary-list__value"
    >
      Value
    </dd>
  </div>
  <div
    class="nhsuk-summary-list__row"
  >
    <dt
      class="nhsuk-summary-list__key"
    >
      Label
    </dt>
    <dd
      class="nhsuk-summary-list__value"
    >
      Value
    </dd>
  </div>
</dl>
`;

exports[`SummaryList > should match snapshot 1`] = `
<dl
  class="nhsuk-summary-list"
  data-testid="summary-list"
>
  <div
    class="nhsuk-summary-list__row"
  >
    <dt
      class="nhsuk-summary-list__key"
    >
      Label
    </dt>
    <dd
      class="nhsuk-summary-list__value"
    >
      Value
    </dd>
    <dd
      class="nhsuk-summary-list__actions"
    >
      <a
        href="#/change"
      >
        Change
      </a>
    </dd>
  </div>
  <div
    class="nhsuk-summary-list__row"
  >
    <dt
      class="nhsuk-summary-list__key"
    >
      Label
    </dt>
    <dd
      class="nhsuk-summary-list__value"
    >
      Value
    </dd>
    <dd
      class="nhsuk-summary-list__actions"
    >
      <a
        href="#/change"
      >
        Change
      </a>
    </dd>
  </div>
  <div
    class="nhsuk-summary-list__row"
  >
    <dt
      class="nhsuk-summary-list__key"
    >
      Label
    </dt>
    <dd
      class="nhsuk-summary-list__value"
    >
      Value
    </dd>
    <dd
      class="nhsuk-summary-list__actions"
    >
      <a
        href="#/change"
      >
        Change
      </a>
    </dd>
  </div>
</dl>
`;
